internal/pkg/cli/deploy/env.go

```diff
@@ -118,15 +132,38 @@ func NewEnvDeployer(in *NewEnvDeployerInput) (*envDeployer, error) {
 	if err != nil {
 		return nil, fmt.Errorf("initialize env describer: %w", err)
 	}
+	ws, err := workspace.Use(afero.NewOsFs())
+	if err != nil {
+		return nil, err
+	}
 	cfnClient := deploycfn.New(envManagerSession, deploycfn.WithProgressTracker(os.Stderr))
-	return &envDeployer{
+
+	var addons addons
+	if in.EnvAddonsFeatureFlag {
+		addonsStack, err := addon.ParseFromEnv(ws)
```

Comment:
This seems OK just a bit unusual to have this done within the constructor 💭 but I think it makes sense given that `UploadArtifacts` and `AddonsTemplate` both require it.

Perhaps an alternative way of writing would be:
```go
type envDeployer struct {
  // existing fields...
  
  parseAddonsOnce sync.Once
  addonsStack  stackBuilder
  addonsErr  error
}


func (d *envDeployer) parseAddons() (stackBuilder, error) {
  d.parseAddonsOnce.Do(func() {
    // parse and assign to fields.
  })
  return d.addonsStack, d.addonsErr
}
```
The main difference is that the parsing-cost is only incurred if parsing addons is necessary. However, if you deem all methods need the addon template then keeping it in the constructor like you did is better.

Reply:
sgtm - will address in a PR that follows https://github.com/aws/copilot-cli/pull/4264 !